### PR TITLE
blui-2338 code coverage

### PR DESCRIPTION
**Pull request by jeffvg**
Fixes # BLUI-2338.

#### Changes proposed in this Pull Request:
- add and config istanbul-instrumenter-loader
- add cypress test framework in example workflow
- add cypress to ci config
- add 37 additional test specs
- add clover reporting for cypress
- add flags for unit_tests and ui_tests
- use ngx-build-plus:dev-server for local dev hosting
- update validation doc
- add several test id's
- comment out randomFailure = () => Math.random() 

branch report - https://app.codecov.io/gh/brightlayer-ui/angular-workflows/tree/feature%2Fblui-2338-cypress-codecov/login-workflow

image

#### To Test:
- clone repo
- cd login-worflow
- run yarn cy:stage && yarn cytest

### CODE REVIEW

**Comment by jeffvg** [comment]
Overall the coverage was bumped from 69% to 80%

**Review by huayunh — APPROVED** [review]
I have not much clue when it comes to software testing, but it looks like you are making the magic happen, so here you go, you have my designer thumbs up.

File: login-workflow/src/components/password-field/password-field.component.ts [hunk 1]
```diff
@@ -28,7 +29,7 @@ import { CrossFieldErrorMatcher } from '../../util/matcher';
                 [errorStateMatcher]="matcher"
                 (keydown.enter)="enter.emit()"
             />
-            <button type="button" mat-icon-button matSuffix (click)="togglePasswordVisibility()">
+            <button id="visibilityIcon" type="button" mat-icon-button matSuffix (click)="togglePasswordVisibility()">
```
Comment (emclaug2) [inline login-workflow/src/components/password-field/password-field.component.ts:32]: #blui-visibility-icon

File: login-workflow/src/pages/create-account/steps/eula/eula.component.ts [hunk 2]
```diff
@@ -24,6 +24,7 @@ import { BluiAuthTranslations } from '../../../../translations/auth-translations
             #eulaVC
             *ngIf="eula"
             class="blui-auth-full-height"
+            id="eula-container"
```
Comment (emclaug2) [inline login-workflow/src/pages/create-account/steps/eula/eula.component.ts:27]: #blui-eula-container

File: login-workflow/src/pages/create-account/steps/eula/eula.component.ts [hunk 3]
```diff
@@ -39,6 +40,7 @@ import { BluiAuthTranslations } from '../../../../translations/auth-translations
         </blui-empty-state>
         <div *ngIf="eula" class="blui-eula-confirm-agreement">
             <mat-checkbox
+                id="eula-checkbox"
```
Comment (emclaug2) [inline login-workflow/src/pages/create-account/steps/eula/eula.component.ts:43]: #blui-eula-checkbox

File: login-workflow/src/pages/login/login.component.html [hunk 4]
```diff
@@ -23,8 +23,8 @@
     <div class="mat-body-2" style="font-weight: 400">{{ errorMessage }}</div>
 </ng-template>
 
-<mat-form-field *ngIf="debugMode">
-    <mat-label>Language</mat-label>
+<mat-form-field *ngIf="debugMode" id="bluiLanguageSelect">
```
Comment (emclaug2) [inline login-workflow/src/pages/login/login.component.html:26]: #blui-language-select

**Review by emclaug2 — COMMENTED** [review]
Very thorough testing - my only comments are for elements with IDs placed within the login-workflow src/ folder.  Prefacing them with 'blui-' would be consistent with the rest of the class names / identifiers used.

File: login-workflow/src/components/email-field/email-field.component.ts [hunk 5]
```diff
@@ -38,6 +38,7 @@ import { AuthErrorStateMatcher } from '../../util/matcher';
                 (keyup.enter)="enter.emit()"
             />
             <mat-error
+                id="email-helper-text"
```
Comment (emclaug2) [inline login-workflow/src/components/email-field/email-field.component.ts:41]: #blui-email-helper-text

**Comment by jeffvg** [comment]
@emclaug2 this is updated now...

**Review by emclaug2 — APPROVED** [review]

**Comment by emclaug2** [comment]
> @emclaug2 this is updated now...

Thanks for updating this, looks good.

**Pull request merged**